src/main/kotlin/org/rust/ide/actions/CargoTreeAction.kt

```diff
@@ -0,0 +1,47 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.actions
+
+import com.intellij.openapi.actionSystem.AnAction
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.project.DumbAware
+import org.rust.cargo.project.toolwindow.CargoToolWindow
+import org.rust.cargo.runconfig.hasCargoProject
+import javax.swing.JTree
+
+abstract class CargoTreeAction : AnAction(), DumbAware {
+
+    override fun update(e: AnActionEvent) {
+        super.update(e)
+        e.presentation.isEnabledAndVisible = e.project?.hasCargoProject == true && getTree(e) != null
+    }
+
+    class CollapseAll : CargoTreeAction() {
+        override fun actionPerformed(e: AnActionEvent) {
+            val tree = getTree(e) ?: return
+            var row = tree.rowCount - 1
+            while (row >= 0) {
+                tree.collapseRow(row)
+                row--
+            }
+        }
+    }
+
+    class ExpandAll : CargoTreeAction() {
+        override fun actionPerformed(e: AnActionEvent) {
+            val tree = getTree(e) ?: return
+            var row = 0
+            while (row < tree.rowCount) {
```

Comment:
```suggestion
            for (row in 0 until tree.rowCount) {
                tree.expandRow(row)
            }
```

Reply:
`rowCount` is a property whose value may change after `expandRow` call.